enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthCache.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.server.security.enterprise.auth;
+
+import java.time.Clock;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.Consumer;
+
+import com.googlecode.concurrentlinkedhashmap.ConcurrentLinkedHashMap;
+import org.apache.shiro.cache.Cache;
+import org.apache.shiro.cache.CacheException;
+import org.apache.shiro.cache.CacheManager;
+
+import org.neo4j.helpers.collection.Pair;
+
+import static java.util.stream.Collectors.toList;
+
+public class AuthCache<K, V> implements Cache<K,V>
+{
+    private final ConcurrentLinkedHashMap<K,Pair<Long,V>> cacheMap;
+    private final Clock clock;
+    private final long TTL;
+    private long lastValidation;
+    private final ExecutorService executorService;
+
+    AuthCache( Clock clock, long ttl, long maxCapacity )
+    {
+        this.clock = clock;
+        this.TTL = ttl;
+        this.lastValidation = clock.millis();
+        this.executorService = Executors.newSingleThreadExecutor();
+        cacheMap = new ConcurrentLinkedHashMap.Builder<K,Pair<Long,V>>()
+                .maximumWeightedCapacity( maxCapacity )
+                .build();
+    }
+
+    @Override
+    public V get( K key ) throws CacheException
+    {
+        return get( key, ignored -> {} );
+    }
+
+    V get( K key, final Consumer<Boolean> onComplete ) throws CacheException
+    {
+        if ( needsValidation() )
+        {
```

Comment:
This could be a bit of a resource hog on smaller TTL settings. Maybe you can relax the 5% heuristic since you are checking the TTL of the entry you are getting anyways?

Also you do not want to queue more that one validation job at any time, since in bad cases you could end up with a flood of meaningless validation jobs.